designs/sam-config.md

```diff
@@ -0,0 +1,369 @@
+
+This decorator benefits from the same type checking that some SAM CLI parameters already use.
+
+A custom callback function (`configuration_callback`) (for the click option) that takes in a custom configuration parser (`provider`) will have rules in place, on how the corresponding configuration can be retrieved and what are the parts that the configuration parser has access to read from.
+
+```
+provider = attrs.pop("provider", TomlProvider(rules=DefaultRules(), command="build", section="parameters"))
+attrs["type"] = click.STRING
+saved_callback = attrs.pop("callback", None)
+partial_callback = functools.partial(onfiguration_callback, cmd_name, option_name, env_name, saved_callback, provider)
+attrs["callback"] = partial_callback
+click.option(*param_decls, **attrs)(f)
+
+```
+
+Phases
+------
+
+The design can be built in phases.
+
+* No option to specify configuration file or env name
+* Specify configuration file with an environment variable
+* Read `--env` to make sure we can select an appropriate portion in configuration file.
+
+
+`.samrc` Changes
+----------------
+
+This design emphasizes parameter pass throughs with a configuration file and does not change the core working of the SAM CLI itself. The SAM CLI continues to be working just as it was with efficiency gains in usability.
+
+Security
+--------
+
+*Tip: How does this change impact security? Answer the following
+questions to help answer this question better:*
+
+**What new dependencies (libraries/cli) does this change require?**
+
+toml
+
+**What other Docker container images are you using?**
+
+N/A
+
+**Are you creating a new HTTP endpoint? If so explain how it will be
+created & used**
+
+N/A
+
+**Are you connecting to a remote API? If so explain how is this
+connection secured**
+
+N/A
+
+**Are you reading/writing to a temporary folder? If so, what is this
+used for and when do you clean up?**
+
+N/A. But we do read from a confiuration file thats either at a default location or specified by the user via an environment variable.
+
+**How do you validate new .samrc configuration?**
+
+
+
+What is your Testing Plan (QA)?
+===============================
+
+Goal
+----
+
+Configuration files are tested alongside SAM CLI and are expected to work seamlessly with meaningful error messages to steer users towards using configuration file to manage their app workflows.
+
+Pre-requesites
+--------------
+
+N/A
+
+Test Scenarios/Cases
+--------------------
+
+* Integration tests for every command with identifer based overrides, and command line overrides on existing sam configuration file and custom configuration file through environment variables.
+* Tested to work on all platforms
+
+Expected Results
+----------------
+* Works on all platforms
+* Resolution of parameters follows.
+ * CLI parameters -> Config file parameters
+
+Documentation Changes
+=====================
+
+* Addition of a new `--env` parameter per command
```

Comment:
Also need documentation on how to write the config file. How to specify flags like `--debug` that don't have explicit values. How to verify whether a config file is parsable or not.